lib/services/role.go

```diff
@@ -1668,10 +1668,29 @@ func NewKubernetesClusterLabelMatcher(clustersLabels map[string]string) RoleMatc
 
 // Match matches a Kubernetes cluster labels against a role.
 func (l *kubernetesClusterLabelMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
-	ok, _, err := MatchLabels(role.GetKubernetesLabels(typ), l.clusterLabels)
+	ok, _, err := MatchLabels(l.getKubeLabels(role, typ), l.clusterLabels)
 	return ok, trace.Wrap(err)
 }
 
+// getKubeLabels returns kubernetes_labels based on resource version and role type.
+func (l kubernetesClusterLabelMatcher) getKubeLabels(role types.Role, typ types.RoleConditionType) types.Labels {
+	labels := role.GetKubernetesLabels(typ)
+
+	// After the introduction of https://github.com/gravitational/teleport/pull/9759 the
+	// kubernetes_labels started to be respected. Former role behavior evaluated deny rules
+	// even if the kubernetes_labels was empty. To preserve this behavior after respecting kubernetes label the label
+	// logic needs to be aligned.
+	// Default wildcard rules should be added to old V3 and V4 roles deny.kubernetes_labels if
+	// deny.kubernetes_labels is empty to ensure that deny rule will be evaluated
+	// even if kubernetes_labels are empty.
+	//
+	// Backward compatibility logic applies only to V3,V4 Role Deny rules.
+	if len(labels) == 0 && typ == types.Deny && (role.GetVersion() == types.V3 || role.GetVersion() == types.V4) {
```

Comment:
@smallinsky I'm actually thinking - why do we need to check the role version here? If/when we introduce v5, won't we run into the same problem?

Reply:
Cause deny rules always evaluate greedily so I don't think we want users to have to explicitly specify kubernetes_labels in them going forward (that would also be backwards incompatible). Am I missing something?